Tests/LeafKitTests/LeafKitTests.swift

```diff
@@ -917,23 +823,170 @@ final class LeafKitTests: XCTestCase {
 
         XCTAssertEqual(view.string, "Hello barvapor")
     }
+    
+    func testCyclicalError() {
+        var test = TestFiles()
+        test.files["/a.leaf"] = "#extend(\"b\")"
+        test.files["/b.leaf"] = "#extend(\"c\")"
+        test.files["/c.leaf"] = "#extend(\"a\")"
+        
+        let renderer = LeafRenderer(
+            configuration: .init(rootDirectory: "/"),
+            cache: DefaultLeafCache(),
+            files: test,
+            eventLoop: EmbeddedEventLoop()
+        )
+        
+        do {
+            _ = try renderer.render(path: "a", context: [:]).wait()
+            XCTFail("Should have thrown LeafError.cyclicalReference")
+        } catch {
+            guard let e = error as? LeafError else { XCTFail("Wrong error: \(error.localizedDescription)"); return }
+            // list of chain references may be out of order
+            XCTAssertEqual(e.localizedDescription, LeafError.cyclicalReference("a", "a -> b -> c -> a").localizedDescription)
+        }
+    }
+    
+    func testDependencyError() {
+        var test = TestFiles()
+        test.files["/a.leaf"] = "#extend(\"b\")"
+        test.files["/b.leaf"] = "#extend(\"c\")"
+        test.files["/c.leaf"] = "#extend(\"d\")"
+        
+        let renderer = LeafRenderer(
+            configuration: .init(rootDirectory: "/"),
+            cache: DefaultLeafCache(),
+            files: test,
+            eventLoop: EmbeddedEventLoop()
+        )
+        
+        do {
+            _ = try renderer.render(path: "a", context: [:]).wait()
+            XCTFail("Should have thrown LeafError.cyclicalReference")
+        } catch {
+            guard let e = error as? LeafError else { XCTFail("Wrong error: \(error.localizedDescription)"); return }
+            // list of chain references may be out of order
```

Comment:
Represent the chain of references as an `Array` rather than a `String` and validate that it is indeed strictly ordered; if it isn't, it doesn't correctly represent the order of references that caused the failure.

Reply:
The chain should inherently be strictly ordered by call reference since it's recursive during the fetch cycle, so F extends A extends Z extends F would be strictly ordered [F,A,Z,F]

Reply:
Right - so the comment saying "may be out of order" is at best fallacious 🙂. It should be represented by an `Array`, though - don't "serialize" information, even/especially in errors, before necessary. It's entirely possible that someone might write a wrapper that can use the chain that caused the cycle to break or repair it in order to retry, for example, and they certainly shouldn't have to parse the chain out of a string that's essentially in "localizedDescription" format and as such can change that format at any time without being an API break. (You know, the way people for ages used `(data as NSData).description` to get a hex string for a `Data` object, only to find themselves completely hosed when Apple finally eventually changed the output of `-[NSData description]`, which they had always had the right to do without it being a compatibility issue.)

Reply:
Oh, shoot. I didn't even see the comment I'd left in. Previously I'd had the chain be a Set and obviously that didn't maintain correctness. Just forgot to chop that line out.

But point taken on flattening the chain too quickly. I've modified that error to hold the array of template paths and only flatten on description, so the error itself still holds the array proper. Will wait to push that until I see if any other major changes are needed once other people chime in.